### ISSUE TITLE
How to train on both CPU and GPU

### ISSUE THREAD

**Issue by TheFriedEgg618** [ultralytics/ultralytics#20935]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have a Dual Xeon setup with a RTX 3050 8GB vRam and 64GB of ram.
While training, only GPU is utilized. Is there a way for me to train using all CPU and GPU?

Environment:
- OS: Window11
- Python 3.13.3
- PyTorch 2.7.1+c118
- Torchvision 0.22.1+cu118
- Ultralytics 8.3.149

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TheFriedEgg618, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for a wide range of [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples—many common questions are already answered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug the issue.

If this is a custom training ❓ Question, please share as much detail as possible, including dataset images, training logs, and verify you're following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- Real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions on [Discourse](https://community.ultralytics.com/)
- Community threads on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response. An Ultralytics engineer will review your question and provide further assistance soon! 😊

**Comment by Y-T-G** [comment]
Training uses both CPU and GPU but for different  things. You can increase `workers`. But the CPU can't help if GPU is a bottleneck or vice-versa.

**Comment by TheFriedEgg618** [comment]
> Training uses both CPU and GPU but for different  things. You can increase `workers`. But the CPU can't help if GPU is a bottleneck or vice-versa.

I see, but the while training, the cpus arent being use at all. For my setup how many `workers` should i set?

**Comment by Y-T-G** [comment]
You can set workers equal to the number of logical cores you have

**Comment by TheFriedEgg618** [comment]
> You can set workers equal to the number of logical cores you have

so smt like this:
```python
results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8, device="cpu",workers=24) #for 24 core
```

**Comment by Y-T-G** [comment]
You shouldn't set `device` to CPU because then it wouldn't use the GPU at all

**Comment by TheFriedEgg618** [comment]
> You shouldn't set `device` to CPU because then it wouldn't use the GPU at all

ok thank you so much i'll try that.

also may i ask if i want to train many different model in one go smt like:
```python
model = YOLO("modelA")
results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8, workers=24)

model = YOLO("modelB")
results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8,workers=24)
```

do I have to free the program's memory before hand?

**Comment by Y-T-G** [comment]
You should use CLI if you want to launch multiple training to avoid any conflicts

**Comment by TheFriedEgg618** [comment]
> > You shouldn't set `device` to CPU because then it wouldn't use the GPU at all
> 
> ok thank you so much i'll try that.
> 
> also may i ask if i want to train many different model in one go smt like:
> 
> model = YOLO("modelA")
> results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8, workers=24)
> 
> model = YOLO("modelB")
> results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8,workers=24)
> do I have to free the program's memory before hand?

I have tried setting `workers` to 48 (i have 48 logical processors), but the training wont even start, most of the time it took forever (20min) loading the dataset, which only have 1k pictures, I observed that only one cpu core is doing anything. Here is the code
```python
result = model.train(data="data.yaml", epochs=150, imgsz=1280, batch=32,  patience=10, workers=48)
``` 

I have also tried doing `torch.set_num_threads` but it doesn't help.

**Comment by TheFriedEgg618** [comment]
> > > You shouldn't set `device` to CPU because then it wouldn't use the GPU at all
> > 
> > 
> > ok thank you so much i'll try that.
> > also may i ask if i want to train many different model in one go smt like:
> > model = YOLO("modelA")
> > results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8, workers=24)
> > model = YOLO("modelB")
> > results = model.train(data="data.yaml", epochs=100, imgsz=1280, batch=8,workers=24)
> > do I have to free the program's memory before hand?
> 
> I have tried setting `workers` to 48 (i have 48 logical processors), but the training wont even start, most of the time it took forever (20min) loading the dataset, which only have 1k pictures, I observed that only one cpu core is doing anything. Here is the code
> 
> result = model.train(data="data.yaml", epochs=150, imgsz=1280, batch=32,  patience=10, workers=48)
> I have also tried doing `torch.set_num_threads` but it doesn't help.

for some reason setting `worker=16` works as expected all core working, then i have a gpu bottleneck.

Thank you for the help.

**Comment by Y-T-G** [comment]
How many cores do you have? You can try half the amount.

**Comment by Y-T-G** [comment]
You can also change this to `os.cpu_count()`. But you need to edit the code.

https://github.com/ultralytics/ultralytics/blob/5921074a34c19faa12e5303d313577d5dd053ce5/ultralytics/utils/__init__.py#L41

**Comment by TheFriedEgg618** [comment]
> How many cores do you have? You can try half the amount.

I did same problem

**Comment by TheFriedEgg618** [comment]
> You can also change this to `os.cpu_count()`. But you need to edit the code.
> 
> [ultralytics/ultralytics/utils/__init__.py](https://github.com/ultralytics/ultralytics/blob/5921074a34c19faa12e5303d313577d5dd053ce5/ultralytics/utils/__init__.py#L41)
> 
> Line 41 in [5921074](/ultralytics/ultralytics/commit/5921074a34c19faa12e5303d313577d5dd053ce5)
> 
>  NUM_THREADS = min(8, max(1, os.cpu_count() - 1))  # number of YOLO multiprocessing threads

`os.cpu_count` and `os.process_cpu_count` returns 48

**Comment by glenn-jocher** [comment]
Great that you found `workers=16` works well! Setting `workers` too high can actually hurt performance due to context switching overhead and memory pressure from too many concurrent processes. The sweet spot is usually between the number of physical cores and total logical cores, which sounds like where you landed.

Since you're now GPU-bottlenecked with your RTX 3050, that's actually the ideal scenario - it means your CPU is efficiently feeding data to the GPU. You could also try reducing `batch=32` to `batch=16` or `batch=24` to see if that helps with GPU memory utilization.

**Comment by TheFriedEgg618** [comment]
> Great that you found `workers=16` works well! Setting `workers` too high can actually hurt performance due to context switching overhead and memory pressure from too many concurrent processes. The sweet spot is usually between the number of physical cores and total logical cores, which sounds like where you landed.
> 
> Since you're now GPU-bottlenecked with your RTX 3050, that's actually the ideal scenario - it means your CPU is efficiently feeding data to the GPU. You could also try reducing `batch=32` to `batch=16` or `batch=24` to see if that helps with GPU memory utilization.

I see that `batch=24` works fine for me, I just have to wait for a while for the data to load into memory. But the bottleneck is large, my RTX 3050 only manage 1 epcho every 20min.

Now i want to try cpu only training to see how it goes. But from all of my testing, i only see at most 6 core used.

**Comment by glenn-jocher** [comment]
For CPU-only training you'll need `device="cpu"` in your train command, but this will be significantly slower than your current GPU setup. The 6-core limitation you're seeing is due to the `NUM_THREADS = min(8, max(1, os.cpu_count() - 1))` setting that was mentioned earlier - it's capped at 8 threads maximum. You can modify that line in the utils file to use more cores, but expect CPU training to take many hours per epoch compared to your current 20 minutes on the RTX 3050.

**Comment by TheFriedEgg618** [comment]
> For CPU-only training you'll need `device="cpu"` in your train command, but this will be significantly slower than your current GPU setup. The 6-core limitation you're seeing is due to the `NUM_THREADS = min(8, max(1, os.cpu_count() - 1))` setting that was mentioned earlier - it's capped at 8 threads maximum. You can modify that line in the utils file to use more cores, but expect CPU training to take many hours per epoch compared to your current 20 minutes on the RTX 3050.
Thanks for your help, i can live with 20 minutes per epoch.

**Comment by glenn-jocher** [comment]
You're welcome! That's a good decision - 20 minutes per epoch with your RTX 3050 is reasonable performance, and you've got a well-optimized setup now with `workers=16` and `batch=24`. Enjoy your training! 🚀
